File: solved.ac/CLASS3/2178.py
import sys
input = sys.stdin.readline
from collections import deque

dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]

def bfs():
    q = deque()
    q.append((0, 0))
    visit = [[0] * m for _ in range(n)]
    visit[0][0] = 1
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < m and 0 <= ny < n:
                if visit[ny][nx] == 0 and board[ny][nx] == 1:
                    visit[ny][nx] = visit[y][x] + 1
                    q.append((nx, ny))

    return visit[-1][-1]

n, m = map(int, input().split())
board = []

for _ in range(n):
    line = list(map(int, input().strip()))
    board.append(line)

print(bfs())

File: solved.ac/CLASS3_JS/2178.js
const fs = require("fs");
const inputPath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const stdin = fs
  .readFileSync(inputPath)
  .toString()
  .split("\n")
  .map((line) => line.trim());

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const solution = () => {
  const [n, m] = input().split(" ").map(Number);
  const board = Array.from({ length: n }, () => input().split("").map(Number));

  const bfs = () => {
    const q = [[0, 0, 1]];
    board[0][0] = 0;
    const dx = [-1, 0, 1, 0];
    const dy = [0, -1, 0, 1];

    while (q.length) {
      const [x, y, cnt] = q.shift();
      for (let i = 0; i < 4; i++) {
        const nx = x + dx[i];
        const ny = y + dy[i];
        if (nx < 0 || nx >= m || ny < 0 || ny >= n) continue;
        if (nx === m - 1 && ny === n - 1) return cnt + 1;
        if (board[ny][nx] === 1) {
          board[ny][nx] = 0;
          q.push([nx, ny, cnt + 1]);
        }
      }
    }
  };
  console.log(bfs());
};

solution();
